Feat!: support Unicode in sqlite, mysql, tsql, postgres, oracle

pruzko opened the pull request:
Added support for `UNICODE` for #4233 

SQLGlot already supports `exp.Chr` that converts integers to chars. It would be useful to have an inverse function that  returns an integer representations of chars.

Here I propose unicode points as the default abstraction since they cover a wide range of symbols, are ASCII compatible, and are pretty much the industry standard. All databases have some function like this, be it `ASCII(x)`, `UNICODE(x)`, or `ORD(x)`. The behavior varies a bit though. SQLite, TSQL, and Postgres return unicode points by default. Oracle's and MySQL's results are dependent on the encoding, hence they require a conversion to UTF.

The documentation is a bit of a mess, especially around the encoding requirements, but I tried this on all of the aforementioned engines and the behavior is consistent.

SQLite: [unicode](https://www.sqlite.org/lang_corefunc.html#unicode)
TSQL: [unicode](https://learn.microsoft.com/en-us/sql/t-sql/functions/unicode-transact-sql?view=sql-server-ver16)
Postgres: [ascii](https://learn.microsoft.com/en-us/sql/t-sql/functions/unicode-transact-sql?view=sql-server-ver16) (returns unicode points for utf8 strings)
MySQL: [ord](https://dev.mysql.com/doc/refman/8.4/en/string-functions.html#function_ord) is encoding dependent, so a [utf32](https://dev.mysql.com/doc/refman/8.4/en/charset-unicode-utf32.html) conversion is necessary
Oracle: [ascii](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/ASCII.html) also needs a conversion through [unistr](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/UNISTR.html)

VaggelisD reviewed (commented):
Hey @pruzko, thank you a lot for these PRs! Will go ahead and look at each one. 

As a general rule, would you mind adding in the description any links/documentation that validate the decisions you've made for each dialect?

Do note that this PR only implements parsing for SQLite (implicitly done through the `exp.Unicode`) and generation for the other dialects, but not the other way around; This means that users won't be able to convert Postgres's `ASCII(x)` to SQLite's `UNICODE(x)` because without Postgres parsing support, `ASCII(x)` will be an `exp.Anonymous` function.

pruzko commented:
Hi, I missed your comments. Sure, I'll address this soon.

pruzko commented:
I cleaned up the code and updated the description. I've also added "ASCII" to Postres's parser.

However, I'd also like to add parsing for MySQL (and Oracle) that would:
1) parse `ord`
2) check if the first argument passed to `ord` is a conversion to utf8
3) if yes, parse it into `exp.Unicode` and **consume both ord and the conversion**, else parse per normal

Do you already have something like this in the codebase please?

georgesittas commented on sqlglot/expressions.py:
```diff
@@ -6599,6 +6599,10 @@ class Unhex(Func):
     pass
 
 
+class Unicode(Func):
+    pass
```
Nvm I checked other dialects, this is g2g.
